src/plotly/api.nim

```diff
@@ -54,14 +87,52 @@ type
     yaxis*: Axis
     yaxis2*: Axis
 
-proc empty(c:Color): bool =
+func newErrorBar*[T: SomeNumber](err: T, color: Color, thickness = 0.0, width = 0.0, visible = true, percent = false): ErrorBar[T] =
+  ## creates an `ErrorBar` object of type `ebkConstantSym` or `ebkPercentSym`, if the `percent` flag
+  ## is set to `true`
+  # NOTE: there is a lot of visual noise in the creation here... change how?
+  if percent == false:
+    result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkConstantSym)
+    result.value = err
+  else:
+    result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkPercentSym)
+    result.percent = err
+
+func newErrorBar*[T: SomeNumber](err: tuple[p, m: T], color: Color, thickness = 0.0, width = 0.0, visible = true, percent = false): ErrorBar[T] =
+  ## creates an `ErrorBar` object of type `ebkConstantAsym`, constant plus and minus errors given as tuple
+  ## or `ebkPercentAsym` of `percent` flag is set to true
+  if percent == false:
+    result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkConstantAsym)
+    result.valuePlus  = err.p
+    result.valueMinus = err.m
+  else:
+    result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkPercentAsym)
+    result.percentPlus  = err.p
+    result.percentMinus = err.m
+
+func newErrorBar*[T: SomeNumber](color: Color, thickness = 0.0, width = 0.0, visible = true): ErrorBar[T] =
+  ## creates an `ErrorBar` object of type `ebkSqrt`
+  result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkSqrt)
+
+func newErrorBar*[T](err: seq[T], color: Color, thickness = 0.0, width = 0.0, visible = true): ErrorBar[T] =
+  ## creates an `ErrorBar` object of type `ebkArraySym`
+  result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkArraySym)
+  result.errors = err
+
+func newErrorBar*[T: SomeNumber](err: tuple[p, m: seq[T]], color: Color, thickness = 0.0, width = 0.0, visible = true): ErrorBar[T] =
+  ## creates an `ErrorBar` object of type `ebkArrayAsym`
+  result = ErrorBar[T](visible: visible, color: color, thickness: thickness, width: width, kind: ebkArraySym)
+  result.errorsPlus  = err.p
+  result.errorsMinus = err.m
+
+func empty(c: Color): bool =
   # TODO: this is also black, but should never need black with alpha == 0
-  return c.r == 0 and c.g == 0 and c.b == 0 and c.a == 0
+  result = c.r == 0 and c.g == 0 and c.b == 0 and c.a == 0
 
-proc `%`*(c:Color): string =
-  return c.toHtmlHex()
+func `%`*(c: Color): string =
```

Comment:
what's the advantage of `func`?

Reply:
Well, a `func` is simply a `proc` with the `{.noSideEffects.}` pragma attached, meaning it should not e.g. change a global variable, perform writes to a file or stdout, mutate one of its arguments, etc. Thus, it is an indication it's much more a mathematical function in the sense that its output only depends on its arguments and nothing more.

It makes the code somewhat easier to reason about to see that something is a pure function.